salt/modules/vagrant.py

```diff
@@ -0,0 +1,628 @@
+# -*- coding: utf-8 -*-
+'''
+Work with virtual machines managed by Vagrant.
+
+.. versionadded:: Oxygen
+
+Mapping between a Salt node id and the Vagrant machine name
+(and the path to the Vagrantfile where it is defined)
+is stored in a Salt sdb database on the Vagrant host (minion) machine.
+In order to use this module, sdb must be configured. An SQLite
+database is the recommended storage method.  The URI used for
+the sdb lookup is "sdb://vagrant_sdb_data".
+
+requirements:
+   - the VM host machine must have salt-minion, Vagrant and a vm provider installed.
+   - the VM host must have a valid definition for `sdb://vagrant_sdb_data`
+
+    Configuration example:
+
+    .. code-block:: yaml
+
+        # file /etc/salt/minion.d/vagrant_sdb.conf
+        vagrant_sdb_data:
+          driver: sqlite3
+          database: /var/cache/salt/vagrant.sqlite
+          table: sdb
+          create_table: True
+
+'''
+
+# Import python libs
+from __future__ import absolute_import, print_function
+import logging
+import os
+
+# Import salt libs
+import salt.utils
+import salt.utils.sdb
+from salt.exceptions import CommandExecutionError, SaltInvocationError
+import salt.ext.six as six
+
+if six.PY3:
+    import ipaddress
+else:
+    import salt.ext.ipaddress as ipaddress
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'vagrant'
+
+VAGRANT_SDB_URL = 'sdb://vagrant_sdb_data/'
+
+
+def __virtual__():
+    '''
+    run Vagrant commands if possible
+    '''
+    # noinspection PyUnresolvedReferences
+    if salt.utils.path.which('vagrant') is None:
+        return False, 'The vagrant module could not be loaded: vagrant command not found'
+    return __virtualname__
+
+
+def _build_sdb_uri(key):
+    '''
+    returns string used to fetch data for "key" from the sdb store.
+
+    Salt node id's are used as the key for vm_ dicts.
+
+    '''
+    return '{}{}'.format(VAGRANT_SDB_URL, key)
+
+
+def _build_machine_uri(machine, cwd):
+    '''
+    returns string used to fetch id names from the sdb store.
+
+    the cwd and machine name are concatenated with '?' which should
+    never collide with a Salt node id -- which is important since we
+    will be storing both in the same table.
+    '''
+    key = '{}?{}'.format(machine, os.path.abspath(cwd))
+    return _build_sdb_uri(key)
+
+
+def _update_vm_info(name, vm_):
+    ''' store the vm_ information keyed by name '''
+    salt.utils.sdb.sdb_set(_build_sdb_uri(name), vm_, __opts__)
+
+    # store machine-to-name mapping, too
+    if vm_['machine']:
+        salt.utils.sdb.sdb_set(
```

Comment:
can you use `__utils__['sdb.sdb_set']` instead of importing the salt.utils.sdb module?

same for the rest of the places you are using salt.utils.

Reply:
Yes, I will revert that.  I thought that the lazy loader was causing my Python3 errors, but I was wrong.